File: app/src/main/java/com/nfcalarmclock/NacTextToSpeech.java
package com.nfcalarmclock;

import android.content.Context;
import android.media.AudioManager;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import java.util.Locale;

/**
 * Text to speech.
 */
public class NacTextToSpeech
	implements TextToSpeech.OnInitListener,
		AudioManager.OnAudioFocusChangeListener
{

	/**
	 * On speaking listener.
	 */
	public interface OnSpeakingListener
	{

		/**
		 * Called when speech engine is done speaking.
		 */
		public void onDoneSpeaking();

		/**
		 * Called when speech engine has started speaking.
		 */
		public void onStartSpeaking();
	}

	/**
	 */
	public static class NacUtteranceListener
		extends UtteranceProgressListener
	{

		/**
		 * Context.
		 */
		private Context mContext;

		/**
		 * Audio focus change listener.
		 */
		private AudioManager.OnAudioFocusChangeListener
			mOnAudioFocusChangeListener;

		/**
		 * On speaking listener.
		 */
		private OnSpeakingListener mOnSpeakingListener;

		/**
		 */
		public NacUtteranceListener(Context context,
			AudioManager.OnAudioFocusChangeListener listener)
		{
			this.mContext = context;
			this.mOnAudioFocusChangeListener = listener;
		}

		/**
		 * @return The context.
		 */
		private Context getContext()
		{
			return this.mContext;
		}

		/**
		 * @return This audio focus change listener.
		 */
		private AudioManager.OnAudioFocusChangeListener
			getOnAudioFocusChangeListener()
		{
			return this.mOnAudioFocusChangeListener;
		}

		/**
		 * @return The on speaking listener.
		 */
		private OnSpeakingListener getOnSpeakingListener()
		{
			return this.mOnSpeakingListener;
		}

		/**
		 */
		@Override
		public void onDone(String utteranceId)
		{
			NacAudio.abandonAudioFocus(this.getContext(),
				this.getOnAudioFocusChangeListener());

			if (this.getOnSpeakingListener() != null)
			{
				this.getOnSpeakingListener().onDoneSpeaking();
			}
		}

		/**
		 */
		@Override
		public void onStart(String utteranceId)
		{
			if (this.getOnSpeakingListener() != null)
			{
				this.getOnSpeakingListener().onStartSpeaking();
			}
		}

		/**
		 */
		@SuppressWarnings("deprecation")
		@Override
		public void onError(String utteranceId)
		{
			NacUtility.printf("onError! %s", utteranceId);
		}

		/**
		 * Set on speaking listener.
		 */
		public void setOnSpeakingListener(OnSpeakingListener listener)
		{
			this.mOnSpeakingListener = listener;
		}

	}

	/**
	 * The context.
	 */
	private Context mContext;

	/**
	 * The speech engine.
	 */
	private TextToSpeech mSpeech;

	/**
	 * Message buffer to speak.
	 */
	private String mBuffer;

	/**
	 * Check if speech engine is initialized.
	 */
	private boolean mInitialized;

	/**
	 * The utterance listener.
	 */
	private NacUtteranceListener mUtterance;

	/**
	 * Audio attributes.
	 */
	private NacAudio.Attributes mAudioAttributes;

	/**
	 */
	public NacTextToSpeech(Context context, OnSpeakingListener listener)
	{
		this.mContext = context;
		this.mSpeech = null;
		this.mBuffer = "";
		this.mInitialized = false;
		this.mUtterance = new NacUtteranceListener(context, this);
		this.mAudioAttributes = new NacAudio.Attributes("Music");

		this.setOnSpeakingListener(listener);
	}

	/**
	 * @return The speech buffer.
	 */
	private String getBuffer()
	{
		return this.mBuffer;
	}

	/**
	 * @return The context.
	 */
	private Context getContext()
	{
		return this.mContext;
	}

	/**
	 * @return The speech engine.
	 */
	private TextToSpeech getTextToSpeech()
	{
		return this.mSpeech;
	}

	/**
	 * @return The utterance listener.
	 */
	private NacUtteranceListener getUtteranceListener()
	{
		return this.mUtterance;
	}

	/**
	 * @return True if there is a buffer present and False otherwise.
	 */
	public boolean hasBuffer()
	{
		String buffer = this.getBuffer();

		return ((buffer != null) && (!buffer.isEmpty()));
	}

	/**
	 * @return True if the speech engine is initialized and False otherwise.
	 */
	public boolean isInitialized()
	{
		return (this.mInitialized && (this.mSpeech != null));
	}

	/**
	 * @return True if the speech engine is running, and False otherwise.
	 */
	public boolean isSpeaking()
	{
		TextToSpeech speech = this.getTextToSpeech();

		return (this.isInitialized() && speech.isSpeaking());
	}

	/**
	 * Change media state when audio focus changes.
	 */
	@Override
	public void onAudioFocusChange(int focusChange)
	{
		//NacUtility.printf("onAudioFocusChange! %d", focusChange);
		String change = "UNKOWN";

		if (focusChange == AudioManager.AUDIOFOCUS_GAIN)
		{
			change = "GAIN";
		}
		else if (focusChange == AudioManager.AUDIOFOCUS_LOSS)
		{
			change = "LOSS";
		}
		else if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT)
		{
			change = "LOSS_TRANSIENT";
		}
		else if (focusChange == AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK)
		{
			change = "LOSS_TRANSIENT_CAN_DUCK";
		}

		//NacUtility.printf("NacTextToSpeech : onAudioFocusChange : AUDIOFOCUS_%s",
		//	change);
	}

	/**
	 */
	@Override
	public void onInit(int status)
	{
		this.mInitialized = (status == TextToSpeech.SUCCESS);

		if (this.isInitialized())
		{
			TextToSpeech speech = this.getTextToSpeech();

			//speech.setLanguage(Locale.US);
			this.mSpeech.setLanguage(Locale.US);

			if (this.hasBuffer())
			{
				String buffer = this.getBuffer();

				this.speak(buffer);
				this.setBuffer("");
			}
		}
	}

	/**
	 * Set the speech message buffer.
	 */
	private void setBuffer(String buffer)
	{
		NacUtility.printf("setting Buffer! '%s'", buffer);
		this.mBuffer = buffer;
	}

	/**
	 * Set on speaking listener.
	 */
	public void setOnSpeakingListener(OnSpeakingListener listener)
	{
		this.getUtteranceListener().setOnSpeakingListener(listener);
	}

	/**
	 * Shutdown the speech engine.
	 */
	public void shutdown()
	{
		//this.stop();

		TextToSpeech speech = this.getTextToSpeech();

		if (speech != null)
		{
			//speech.shutdown();
			this.mSpeech.shutdown();

			this.mSpeech = null;
		}
	}

	/**
	 * @see speak
	 */
	public void speak(String message)
	{
		this.speak(message, this.mAudioAttributes);
	}

	/**
	 * Speak the given text.
	 */
	public void speak(String message, NacAudio.Attributes attrs)
	{
		Context context = this.getContext();
		TextToSpeech speech = this.getTextToSpeech();
		this.mAudioAttributes = attrs;
		NacUtility.printf("Speaking! %s", message);

		if (speech == null)
		{
			this.mSpeech = new TextToSpeech(context, this);

			this.mSpeech.setOnUtteranceProgressListener(this.mUtterance);
		}

		if (this.isInitialized())
		{
			if(!NacAudio.requestAudioFocusTransient(context, this, attrs))
			{
				NacUtility.printf("Audio Focus TRANSIENT NOT Granted!");
				return;
			}

			//speech.speak(message, TextToSpeech.QUEUE_FLUSH, null, "AlarmTime");
			this.mSpeech.speak(message, TextToSpeech.QUEUE_FLUSH, null, "AlarmTime");
		}
		else
		{
			this.setBuffer(message);
		}
	}

	/**
	 * Stop the speech engine.
	 */
	public void stop()
	{
		TextToSpeech speech = this.getTextToSpeech();

		if (speech != null)
		{
			//speech.stop();
			this.mSpeech.stop();
		}
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/alarm/options/tts/NacTextToSpeech.kt
package com.nfcalarmclock.alarm.options.tts

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioManager
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import com.nfcalarmclock.R
import com.nfcalarmclock.log.NacLog
import com.nfcalarmclock.system.media.NacAudioAttributes
import com.nfcalarmclock.system.media.requestFocusGainTransient
import com.nfcalarmclock.system.toBundle
import com.nfcalarmclock.view.quickToast
import java.util.Locale

/**
 * Text to speech.
 *
 * @param context Context.
 * @param onStartSpeaking Function to call when starting to speak.
 * @param onDoneSpeaking Function to call when done speaking.
 * @param onErrorSpeaking Function to call when an error occurred trying to speak.
 */
class NacTextToSpeech(
	context: Context,
	onInit: (TextToSpeech, Int) -> Unit = { _, _ -> },
	onStartSpeaking: () -> Unit = {},
	onDoneSpeaking: () -> Unit = {},
	onErrorSpeaking: () -> Unit = {
		quickToast(context, R.string.error_message_text_to_speech_audio_focus)
	},
	private val utteranceId: String = UTTERANCE_ID
)
{

	/**
	 * Utterance listener.
	 */
	class NacUtteranceListener(
		val onStartSpeaking: () -> Unit = {},
		val onDoneSpeaking: () -> Unit = {},
		val onErrorSpeaking: () -> Unit = {}
	) : UtteranceProgressListener()
	{

		/**
		 * Done speaking.
		 */
		override fun onDone(utteranceId: String)
		{
			onDoneSpeaking()
		}

		/**
		 * Starting to speak.
		 */
		override fun onStart(utteranceId: String)
		{
			onStartSpeaking()
		}

		/**
		 * Error speaking.
		 */
		@Deprecated("Deprecated in Java")
		override fun onError(utteranceId: String)
		{
			NacLog.e("On speaking error. utteranceId=$utteranceId")
			onErrorSpeaking()
		}

		/**
		 * Error speaking.
		 */
		override fun onError(utteranceId: String, errorCode: Int)
		{
			NacLog.e("On speaking error. utteranceId=$utteranceId | errorCode=$errorCode")
			onErrorSpeaking()
		}

	}

	/**
	 * Audio manager.
	 */
	val audioManager: AudioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

	/**
	 * The speech engine.
	 *
	 * The TextToSpeech.OnInitListener is the lambda.
	 */
	val textToSpeech: TextToSpeech = TextToSpeech(context) { status ->

		// Cleanup has been called before init is finished
		if (isCleanedUp)
		{
			return@TextToSpeech
		}

		// Set the initialization status
		isInitialized = (status == TextToSpeech.SUCCESS)

		// Initialization was a succes and there is a message in the buffer
		if (isInitialized && hasBuffer())
		{
			// Say what is in the buffer
			speak(bufferMessage, bufferAudioAttributes!!)
		}

		// Call the listener
		onInit(textToSpeech, status)

	}

	/**
	 * Message to buffer and speak when ready.
	 */
	private var bufferMessage: String = ""

	/**
	 * Audio attributes to buffer and use when ready.
	 */
	private var bufferAudioAttributes: NacAudioAttributes? = null

	/**
	 * Check if speech engine is initialized.
	 */
	var isInitialized: Boolean = false

	/**
	 * Whether the TTS resources have been cleaned up or not.
	 */
	var isCleanedUp: Boolean = false

	/**
	 * The utterance listener.
	 */
	private val utteranceListener: NacUtteranceListener = NacUtteranceListener(
		onStartSpeaking = onStartSpeaking,
		onDoneSpeaking = onDoneSpeaking,
		onErrorSpeaking = onErrorSpeaking
	)

	/**
	 * Constructor.
	 */
	init
	{
		// Setup text to speech listener
		textToSpeech.setOnUtteranceProgressListener(utteranceListener)
	}

	/**
	 * Cleanup any used resources.
	 */
	fun cleanup()
	{
		// Shutdown the TTS engine
		textToSpeech.shutdown()

		// Set the cleanup flag
		isCleanedUp = true
	}

	/**
	 * Clear the buffered message and audio attributes.
	 */
	private fun clearBuffer()
	{
		bufferMessage = ""
		bufferAudioAttributes = null
	}

	/**
	 * Check if there is a buffer present.
	 *
	 * @return True if there is a buffer present, and False otherwise.
	 */
	fun hasBuffer() : Boolean
	{
		return bufferMessage.isNotEmpty()
	}

	/**
	 * Check if the speech engine is running.
	 *
	 * @return True if the speech engine is running, and False otherwise. Will
	 *         also return True if the IllegalArgumentException is raised.
	 */
	fun isSpeaking() : Boolean
	{
		return try
		{
			this.isInitialized && textToSpeech.isSpeaking
		}
		catch (e: IllegalArgumentException)
		{
			NacLog.e("Unable to check whether TTS is speaking", throwable = e)
			true
		}
	}

	/**
	 * Speak the given text.
	 */
	fun speak(message: String, attrs: NacAudioAttributes)
	{
		// Cleanup has been called before init is finished
		if (isCleanedUp)
		{
			return
		}

		// TTS object is already initialized
		if (isInitialized)
		{
			// Gain transient audio focus
			if (!audioManager.requestFocusGainTransient(null, attrs))
			{
				// Clear the buffer, just in case
				clearBuffer()

				// Call error and done listeners
				utteranceListener.onErrorSpeaking()
				utteranceListener.onDoneSpeaking()
				return
			}

			// Set the language
			textToSpeech.language = Locale.getDefault()

			// Set the speech rate
			if (attrs.speechRate != 0f)
			{
				textToSpeech.setSpeechRate(attrs.speechRate)
			}

			// Set the voice
			if (attrs.voice.isNotEmpty())
			{
				textToSpeech.voices
					.find { it.name == attrs.voice }
					?.let { textToSpeech.voice = it }
			}

			// Create TTS audio attributes and bundle any TTS params
			val audioAttributes = AudioAttributes.Builder()
				.setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
				.setUsage(attrs.audioUsage)
				.build()
			val bundle = attrs.toBundle()

			// Speak the message
			textToSpeech.setAudioAttributes(audioAttributes)
			textToSpeech.speak(message, TextToSpeech.QUEUE_FLUSH, bundle, utteranceId)

			// Clear the buffer
			clearBuffer()
		}
		// Buffer the message and audio attributes
		else
		{
			bufferMessage = message
			bufferAudioAttributes = attrs
		}
	}

	companion object
	{

		/**
		 * Utterance ID when speaking through the TTS engine.
		 */
		const val UTTERANCE_ID = "NacAlarmTts"

	}

}